src/main/java/graphql/EngineRunningState.java

```diff
@@ -0,0 +1,209 @@
+package graphql;
+
+import graphql.execution.EngineRunningObserver;
+import graphql.execution.ExecutionId;
+import org.jspecify.annotations.Nullable;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static graphql.Assert.assertTrue;
+import static graphql.execution.EngineRunningObserver.RunningState.NOT_RUNNING;
+import static graphql.execution.EngineRunningObserver.RunningState.RUNNING;
+
+@Internal
+public class EngineRunningState {
+
+    @Nullable
+    private final EngineRunningObserver engineRunningObserver;
+    @Nullable
+    private final GraphQLContext graphQLContext;
+    @Nullable
+    private volatile ExecutionId executionId;
+
+    private final AtomicInteger isRunning = new AtomicInteger(0);
+
+    @VisibleForTesting
+    public EngineRunningState() {
+        this.engineRunningObserver = null;
+        this.graphQLContext = null;
+        this.executionId = null;
+    }
+
+    public EngineRunningState(ExecutionInput executionInput) {
+        EngineRunningObserver engineRunningObserver = executionInput.getGraphQLContext().get(EngineRunningObserver.ENGINE_RUNNING_OBSERVER_KEY);
```

Comment:
Rather than `ExecutionInput` it should just be `GraphqlContext` right.  You dont need the ExecutionInput do you ?

Reply:
Ahh you do - ignore me